util/test/demos/apple/apple_window.cpp

```diff
@@ -0,0 +1,192 @@
+#define CA_PRIVATE_IMPLEMENTATION
+#include "official/metal-cpp.h"
+
+static NS::Application *pSharedApplication = NULL;
+MyAppDelegate *AppleWindow::pAppDelegate = NULL;
+
+class MyAppDelegate : public NS::ApplicationDelegate
+{
+public:
+  ~MyAppDelegate();
+
+  NS::Menu *createMenuBar();
+
+  virtual void applicationWillFinishLaunching(NS::Notification *pNotification) override;
+  virtual void applicationDidFinishLaunching(NS::Notification *pNotification) override;
+  virtual bool applicationShouldTerminateAfterLastWindowClosed(NS::Application *pSender) override;
+
+  void CreateWindow(int width, int height, const char *title);
+  NS::View *GetContentView();
+  NS::Window *GetWindow();
+
+private:
+  NS::Window *_pWindow;
+};
+
+MyAppDelegate::~MyAppDelegate()
+{
+  _pWindow->release();
+}
+
+NS::Menu *MyAppDelegate::createMenuBar()
+{
+  using NS::StringEncoding::UTF8StringEncoding;
+
+  NS::Menu *pMainMenu = NS::Menu::alloc()->init();
+  NS::MenuItem *pAppMenuItem = NS::MenuItem::alloc()->init();
+  NS::Menu *pAppMenu = NS::Menu::alloc()->init(NS::String::string("Appname", UTF8StringEncoding));
+
+  NS::String *appName = NS::RunningApplication::currentApplication()->localizedName();
+  NS::String *quitItemName =
+      NS::String::string("Quit ", UTF8StringEncoding)->stringByAppendingString(appName);
+  SEL quitCb =
+      NS::MenuItem::registerActionCallback("appQuit", [](void *, SEL, const NS::Object *pSender) {
+        auto pApp = NS::Application::sharedApplication();
+        pApp->terminate(pSender);
+      });
+
+  NS::MenuItem *pAppQuitItem =
+      pAppMenu->addItem(quitItemName, quitCb, NS::String::string("q", UTF8StringEncoding));
+  pAppQuitItem->setKeyEquivalentModifierMask(NS::EventModifierFlagCommand);
+  pAppMenuItem->setSubmenu(pAppMenu);
+
+  NS::MenuItem *pWindowMenuItem = NS::MenuItem::alloc()->init();
+  NS::Menu *pWindowMenu = NS::Menu::alloc()->init(NS::String::string("Window", UTF8StringEncoding));
+
+  SEL closeWindowCb =
+      NS::MenuItem::registerActionCallback("windowClose", [](void *, SEL, const NS::Object *) {
+        auto pApp = NS::Application::sharedApplication();
+        pApp->windows()->object<NS::Window>(0)->close();
+      });
+  NS::MenuItem *pCloseWindowItem =
+      pWindowMenu->addItem(NS::String::string("Close Window", UTF8StringEncoding), closeWindowCb,
+                           NS::String::string("w", UTF8StringEncoding));
+  pCloseWindowItem->setKeyEquivalentModifierMask(NS::EventModifierFlagCommand);
+
+  pWindowMenuItem->setSubmenu(pWindowMenu);
+
+  pMainMenu->addItem(pAppMenuItem);
+  pMainMenu->addItem(pWindowMenuItem);
+
+  pAppMenuItem->release();
+  pWindowMenuItem->release();
+  pAppMenu->release();
+  pWindowMenu->release();
+
+  return pMainMenu->autorelease();
+}
+
+void MyAppDelegate::applicationWillFinishLaunching(NS::Notification *pNotification)
+{
+  NS::Menu *pMenu = createMenuBar();
+  NS::Application *pApp = reinterpret_cast<NS::Application *>(pNotification->object());
```

Comment:
Similar to the note above, avoid C++ casts like this, `(NS::Application *)pNotification->object()` is fine here.